File: go/day13/main.go
package main

import (
	input "aoc2020/inpututils"

	"fmt"
	"strconv"
	"strings"
)

func main() {
	fmt.Println("--- Part One ---")
	fmt.Println(Part1("input.txt"))

	fmt.Println("--- Part Two ---")
	fmt.Println(Part2("input.txt"))
}

func Part1(filename string) int {
	lines := input.ReadLines(filename)

	earliestTime := toInt(lines[0])
	buses := lines[1]

	shortestWait, currentBusId := int(^uint(0)>>1), 0

	for _, bus := range strings.Split(buses, ",") {
		if bus == "x" {
			continue
		}
		b := toInt(bus)
		wait := (((earliestTime / b) * b) + b - earliestTime)
		if wait < shortestWait {
			shortestWait = wait
			currentBusId = b
		}
	}
	return shortestWait * currentBusId
}

func Part2(filename string) int {
	lines := input.ReadLines(filename)

	buses := strings.Split(lines[1], ",")
	runningProduct, earliestBus := 1, 0

	for idx, bus := range buses {
		if bus == "x" {
			continue
		}
		for (earliestBus+idx)%toInt(bus) != 0 {
			earliestBus += runningProduct
		}
		runningProduct *= toInt(bus)
	}
	return earliestBus
}

func check(err error) {
	if err != nil {
		panic(err)
	}
}

func toInt(s string) int {
	i, err := strconv.Atoi(s)
	check(err)
	return i
}

File: python/src/day13/main.py
from src.common.file_utils import get_path, read_lines


def part_one(filename: str) -> int:
    lines = read_lines(get_path(__file__, filename))
    earliest, shortest_wait, bus_id = int(lines[0]), float('inf'), None

    for b in [int(x) for x in lines[1].split(',') if x.isnumeric()]:
        wait = (((earliest // b) * b) + b) - earliest
        if wait < shortest_wait:
            shortest_wait, bus_id = wait, b
    
    return shortest_wait * bus_id if shortest_wait else -1


def part_two(filename: str) -> int:
    lines = read_lines(get_path(__file__, filename))
    running_product, earliest_bus = 1, 0

    # uses chinese remainder theorem
    for (index, bus) in enumerate(lines[1].split(",")):
        if bus != "x":
            while ((earliest_bus + index) % int(bus)) != 0:
                earliest_bus += running_product
            running_product *= int(bus)
    
    return earliest_bus


if __name__ == '__main__':

    print("---Part One---")
    part_one_res = part_one("input.txt")
    print(part_one_res)

    print("---Part Two---")
    part_two_res = part_two("input.txt")
    print(part_two_res)
